testhelper/gittest/gittest.go

```diff
@@ -0,0 +1,149 @@
+	gitRoot := filepath.Join(tempDir, org, repo)
+	require.NoErrorf(t, os.MkdirAll(gitRoot, os.ModePerm), "should be able to create %s", gitRoot)
+
+	out, err := execCmd("rsync", "--recursive", source, workingDir)
+	require.NoErrorf(t, err, "should be able to copy %s to %s: %s", source, workingDir, out)
+
+	gitPath, err := exec.LookPath("git")
+	require.NoError(t, err, "should be able to find git in PATH")
+	out, err = execCmd("git", "init", workingDir)
+	require.NoErrorf(t, err, "should be able to initialize git repository: %s", out)
+	out, err = execCmd("git", "-C", workingDir, "branch", "-m", "main")
+	require.NoErrorf(t, err, "should be able to rename the default branch to main: %s", out)
+	out, err = execCmd("git", "-C", workingDir, "add", ".")
+	require.NoErrorf(t, err, "should be able to add files to git repository: %s", out)
+	commitCmd := exec.Command("git", "-C", workingDir, "commit", "-m", "initial commit")
+	commitCmd.Env = append(commitCmd.Env, "GIT_AUTHOR_NAME=git test", "GIT_COMMITTER_NAME=git test", "GIT_AUTHOR_EMAIL=gittest@example.com", "GIT_COMMITTER_EMAIL=gittest@example.com")
+	require.NoError(t, commitCmd.Run(), "should be able to commit files to git repository")
+
+	out, err = execCmd("git", "clone", "--bare", workingDir, gitRoot)
+	require.NoErrorf(t, err, "should be able to clone the bare repository: %s", out)
+
+	handler := &cgi.Handler{
+		Path: gitPath,
+		// Root: org + "/" + repo + "/",
+		Args: []string{"http-backend"},
+		Env: []string{
+			fmt.Sprintf("GIT_PROJECT_ROOT=%s", tempDir),
+			"GIT_HTTP_EXPORT_ALL=true",
+			"REMOTE_USER=git",
+		},
+	}
+
+	localIP := getLocalIP(t)
+	var s *httptest.Server
+	if !secure {
+		s = httptest.NewServer(handler)
+	} else {
+		s = httptest.NewTLSServer(localIP, handler)
+		certPath := filepath.Join(tempDir, "server.crt")
+		require.NoError(t, writeServerCA(s, certPath))
+		t.Setenv("SSL_CERT_FILE", certPath)
+
+	}
+	serverURL, err := url.Parse(s.URL)
+	require.NoError(t, err)
+	_, port, err := net.SplitHostPort(serverURL.Host)
+	require.NoError(t, err)
+	serverURL.Host = net.JoinHostPort(getLocalIP(t), port)
+	url := serverURL.String() + "/" + org + "/" + repo
+	return &Server{
+		Server: s,
+		URL:    url,
+	}
+}
+
+// getServerCA returns a byte slice containing the PEM encoding of the server's CA certificate
+func (s *Server) GetServerCA() []byte {
+	return getServerCA(s.Server)
+}
+
+func getServerCA(server *httptest.Server) []byte {
+	return pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: server.TLS.Certificates[0].Certificate[0]})
+}
+
+// writeServerCA writes the PEM-encoded server certificate to a given path
+func writeServerCA(server *httptest.Server, path string) error {
+	certOut, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer certOut.Close()
+
+	if _, err := certOut.Write(getServerCA(server)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getLocalIP(t testing.TB) string {
+	conn, err := net.Dial("udp", "8.8.8.8:80")
```

Comment:
this seems like a very common way: https://github.com/search?q=%22net.Dial%28%5C%22udp%5C%22%2C+%5C%228.8.8.8%3A80%5C%22%29%22&type=code

Reply:
stackoverflow, that's why!
https://stackoverflow.com/questions/23558425/how-do-i-get-the-local-ip-address-in-go/37382208#37382208